Create Application Layer

Nafanyan opened the pull request:
Поправочки по фаренгейту

Nafanyan committed:
Create Application Layer

SherlockD commented on Application/Events/EventsBindings.cs:
```diff
@@ -0,0 +1,21 @@
+    {
+        public static IServiceCollection AddEventsBindings(this IServiceCollection services)
+        {
+            services.AddScoped<IAddEventCommandHandler, AddEventCommandHandler>();
```
IqueryHandler, ICommandHandler

SherlockD commented on Application/Events/Commands/CreateEvent/AddEventCommandHandler.cs:
```diff
@@ -0,0 +1,32 @@
+            Event e = new Event(addEventCommand.Name, addEventCommand.Description, eventPeriod);
+            _eventRepository.Add(e);
+        }
+        private void CommandValidation(AddEventCommand addEventCommand)
```
Лучше вынести в отдельные классы по single responsibility

SherlockD commented on Application/Events/Commands/CreateEvent/AddEventCommandHandler.cs:
```diff
@@ -0,0 +1,32 @@
+        {
+        }
+
+        public void Execute(AddEventCommand addEventCommand)
```
Handler

SherlockD replied:
Подумай что будет возвращать хендлер

SherlockD commented on Application/Events/BaseEventUseCase.cs:
```diff
@@ -0,0 +1,17 @@
+{
+    public abstract class BaseEventUseCase
+    {
+        protected readonly IEventRepository _eventRepository;
```
Протектид поля с большой буквы

SherlockD commented on Application/Events/BaseEventUseCase.cs:
```diff
@@ -0,0 +1,17 @@
+
+namespace Application.Events
+{
+    public abstract class BaseEventUseCase
```
Base event handler

SherlockD commented on Application/Events/Commands/CreateEvent/AddEventCommandHandler.cs:
```diff
@@ -0,0 +1,32 @@
+            CommandValidation(addEventCommand);
+
+            EventPeriod eventPeriod = new EventPeriod(addEventCommand.StartEvent, addEventCommand.EndEvent);
+            Event e = new Event(addEventCommand.Name, addEventCommand.Description, eventPeriod);
```
Лучше писать полностью

Nafanyan committed:
Создал класс возвращения результата для сущности Event

Nafanyan committed:
Поправил ошибки визуала. Добавил сущность возврата результата для Users.

Nafanyan committed:
Убрал абстрактный класс

Nafanyan committed:
Небольший фикс

Nafanyan committed:
Небольшие правки

Nafanyan committed:
.

SherlockD commented on Application/Users/Commands/CreateUser/AddUserCommandHandler.cs:
```diff
@@ -0,0 +1,29 @@
+
+namespace Application.Users.Commands.CreateUser
+{
+    public class AddUserCommandHandler : IUserCommandHandler<AddUserCommand>
```
ICommandHandler, IQueryHandler, нам не нужно для каждой сущности свою абстракцию

SherlockD commented on Application/Events/Commands/CreateEvent/AddEventCommand.cs:
```diff
@@ -0,0 +1,10 @@
+﻿namespace Application.Events.Commands.CreateEvent
+{
+    public class AddEventCommand
```
По именованию, везде все ок, а тут namespace CreateEvent, команда AddEvent

SherlockD commented on Application/IValidation.cs:
```diff
@@ -0,0 +1,7 @@
+{
+    public interface IValidation<T>
+    {
+        string Validation(T inputData);
```
Не совсем верно указывать строку в качестве результата валидации. Это совсем не прозрачно получается. Предлагаю возвращать составной объект из полей bool IsFail и Error с описанием ошибки. Для детализации ошибок

SherlockD commented on Application/Results/ResultCommand.cs:
```diff
@@ -0,0 +1,12 @@
+﻿namespace Application.Result
+{
+    public class ResultCommand
```
Тут тоже строкой не верное решение передавать результат. По сути команда она не возвращает ничего, она изменяет состояние. Как вариант можно возвращать id измененного или созданного объекта, но это тоже нужно смотреть нужно ли это овобще или нет. Тут также можно хранить объект Error

SherlockD replied:
По именованию: Result Command: Команда результата
                              Command Result: Результат команды

SherlockD commented on Application/Results/ResultQuery.cs:
```diff
@@ -0,0 +1,18 @@
+    public class ResultQuery<T> where T : class
+    {
+        public T ResultObj { get; private set; }
+        public string MSG { get; private set; }
```
Тут в плане MSG то же самое что и в команде

SherlockD commented on Application/Events/Queries/IEventQueryHandler.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace Application.Events.Queries
+{
+    public interface IEventQueryHandler<T, Q> 
```
Что такое T и Q) По правилам именования дженериков у над должно быть TQuery и TResultData к примеру, но тут можно просто TQuery, TResult, Handle будет такого вида Task<TResult> Handle(TQuery query)

SherlockD commented on Application/Users/Commands/CreateUser/AddUserCommandValidation.cs:
```diff
@@ -0,0 +1,30 @@
+
+namespace Application.Users.Commands.CreateUser
+{
+    class AddUserCommandValidation : IValidation<AddUserCommand>
```
Не забывай модификаторы доступа класса обозначать, это для большей читабельности

SherlockD commented on Application/Users/Commands/CreateUser/AddUserCommandValidation.cs:
```diff
@@ -0,0 +1,30 @@
+                return "The login cannot be empty/cannot be null";
+            }
+
+            if (_userRepository.GetAll().Result.Where(u => u.Login == command.Login).
```
Все i/o даже в валидаторах нужно делать async, у тебя будет два валидатора IValidator, IAsyncValidator :)

SherlockD commented on Application/IValidation.cs:
```diff
@@ -0,0 +1,7 @@
+﻿namespace Application
+{
+    public interface IValidation<T>
```
IValidator, сущность которая производит валидацию

SherlockD commented on Application/Users/Commands/CreateUser/AddUserCommand.cs:
```diff
@@ -0,0 +1,8 @@
+{
+    public class AddUserCommand
+    {
+            public string Login { get; init; }
```
отступы

SherlockD commented on Application/Users/Commands/UpdateUserLogin/UpdateUserLoginCommandValidation.cs:
```diff
@@ -0,0 +1,42 @@
+                return "A user with this login already exists";
+            }
+
+            User user = _userRepository.GetById(command.Id).Result;
```
Тут тоже i/o тоже должно быть асинхронным. В целом где есть i/o везде будем использовать асинхронные функции, так что не бойся их юзать даже в валидаторе)

SherlockD commented on Application/Users/Queries/GetEvents/GetEventsQueryHandler.cs:
```diff
@@ -0,0 +1,28 @@
+            _getEventsQueryValidation = new GetEventsQueryValidation(userRepository);
+        }
+
+        public async Task<ResultQuery<IReadOnlyList<Event>>> Handle(GetEventsQuery getEventsQuery)
```
Тут молодец, IReadOnlyList хорошее решение использовать, так как результат бизнес логики мы не в праве изменять на уровень выше

SherlockD commented on Application/Users/Queries/GetEvents/GetEventsQueryHandler.cs:
```diff
@@ -0,0 +1,28 @@
+    public class GetEventsQueryHandler : IUserQueryHandler<IReadOnlyList<Event>, GetEventsQuery>
+    {
+        private readonly IUserRepository _userRepository;
+        private readonly GetEventsQueryValidation _getEventsQueryValidation;
```
Ага, вот что еще заметил, тут мы можем использовать интерфейс валидатора и внедрять через DI конкретный валидатор сюда

SherlockD commented on Application/Users/Queries/GetUserById/GetUserByIdQuery.cs:
```diff
@@ -0,0 +1,8 @@
+    public class GetUserByIdQuery
+    {
+        public long Id { get; init; }
+
```
Пробел, надеюсь ты не злишься на меня :D

SherlockD commented on Application/Users/Queries/GetUsers/GetUsersQuery.cs:
```diff
@@ -0,0 +1,6 @@
+﻿namespace Application.Users.Queries.GetUsers
+{
+    public class GetUsersQuery
```
А где нам на фронте может понадобится жтот метод получения всех пользователей?

SherlockD commented on Application/Users/Queries/IUserQueryHandler.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace Application.Users.Queries
+{
+    public interface IUserQueryHandler<T, Q>
```
Тут тоже что в команде, T,Q не совсем верное именование

SherlockD commented on Application/Users/Queries/GetEvents/GetEventsQueryHandler.cs:
```diff
@@ -0,0 +1,28 @@
+            string msg = _getEventsQueryValidation.Validation(getEventsQuery);
+            if (msg == "Ok")
+            {
+                return new ResultQuery<IReadOnlyList<Event>>(await _userRepository.GetEvents(getEventsQuery.UserId), msg);
```
В таких моментах лучше разделить немного по переменным, для общей читаемости, это не критично, можно сказать просто вот взгляд со стороны.

SherlockD replied:
Не поправил :)

Nafanyan committed:
убрал разделение интерфесов command query для Events и Users

Nafanyan committed:
Сделал, чтобы валидаторы возвращали составной тип

Nafanyan committed:
исправил using

Nafanyan committed:
Поправил оступы и прочие недочеты

Nafanyan committed:
пробел

Nafanyan committed:
добавил асинхронность в валидацию

Nafanyan committed:
Сделал валидаторы в командах и запросах через DI

SherlockD commented on Application/Events/Commands/CreateEvent/CreateEventCommandHandler.cs:
```diff
@@ -0,0 +1,32 @@
+    public class CreateEventCommandHandler : ICommandHandler<CreateEventCommand>
+    {
+        private readonly IEventRepository _eventRepository;
+        private readonly IValidator<CreateEventCommand> _createEventCommandValidation;
```
_createEventCommandValidat**or**

SherlockD commented on Application/Events/Commands/CreateEvent/CreateEventCommandValidation.cs:
```diff
@@ -0,0 +1,43 @@
+        public ValidationResult Validation(CreateEventCommand command)
+        {
+            string error = "No errors";
+            ValidationResult validationResult = new ValidationResult(false, error);
```
Почему бы нам не оставить ошибку пустой, если ее нет :)

SherlockD commented on Application/Validation/ValidationResult.cs:
```diff
@@ -0,0 +1,14 @@
+﻿namespace Application.Validation
+{
+    public class ValidationResult
```
Смотри, тут можем уростить себе жинь немного, если мы в этом классе сделаем два статический метода Ok(), Fail(string error);

public static ValidatrionResult Ok()
{
    return new ValidationResult();
}

public static ValidatrionResult Fail(string error)
{
    return new ValidationResult(error);
}

SherlockD commented on Application/Validation/ValidationResult.cs:
```diff
@@ -0,0 +1,14 @@
+        public bool IsFail { get; private set; }
+        public string Error { get; private set; }
+
+        public ValidationResult(bool isFail, string error)
```
Предлагаю сделать error необязательным параметром:
https://professorweb.ru/my/csharp/charp_theory/level5/5_10.php

А поле isFail не принимать в контрукторе, а вычислеть IsFail => Error != null

SherlockD commented on Application/Results/QueryResult.cs:
```diff
@@ -0,0 +1,21 @@
+
+namespace Application.Result
+{
+    public class QueryResult<T> where T : class
```
TQueryResultData

SherlockD replied:
Обрати внимание везде по проекту, где используешь Generic, о чем я писал в прошлом пулреквесте, просто T,Q признак плохого тона, лучше всего указывать полностью названия того что там будет находится, грубо говоря.

SherlockD commented on Application/CQRSInterfaces/ICommandHandler.cs:
```diff
@@ -0,0 +1,9 @@
+
+namespace Application.Interfaces
+{
+    public interface ICommandHandler<T> where T : class
```
Тут тоже TCommand

SherlockD commented on Application/Events/Queries/GetEvent/GetEventQueryHandler.cs:
```diff
@@ -0,0 +1,32 @@
+
+namespace Application.Events.Queries.GetEvent
+{
+
```
пробел

SherlockD commented on Application/Events/Queries/GetEvent/GetEventQueryHandler.cs:
```diff
@@ -0,0 +1,32 @@
+    public class GetEventQueryHandler : IQueryHandler<Event, GetEventQuery>
+    {
+        private readonly IEventRepository _eventRepository;
+        private IValidator<GetEventQuery> _eventQueryValidation;
```
readonly

SherlockD commented on Application/Events/Queries/GetEvent/GetEventQueryValidation.cs:
```diff
@@ -0,0 +1,63 @@
+            _eventRepository = eventRepository;
+        }
+
+        public ValidationResult Validation(GetEventQuery query)
```
Это все можно запихать в async validator, тут как будто бы не нужно два интерфейса

SherlockD commented on Application/Users/Queries/GetEvents/GetEventsQueryValidation.cs:
```diff
@@ -0,0 +1,44 @@
+            string error = "No errors";
+            ValidationResult validationResult = new ValidationResult(false, error);
+
+            if (await _userRepository.GetById(query.UserId) == null)
```
Вот для таких случаев лучше сделать в репозитории метод Contains

Nafanyan committed:
Исправил последние замечания

Nafanyan committed:
пробелы

SherlockD commented on Application/Events/Commands/UpdateEvent/UpdateEventCommandValidation.cs:
```diff
@@ -0,0 +1,47 @@
+            string error;
+            if (command.StartEvent == null)
+            {
+                error = "The start date cannot be empty/cannot be null";
```
Тут в целом можно прям в Fail() текст ошибки писать, без переменной)

SherlockD commented on Application/Events/Queries/GetEvent/GetEventQueryHandler.cs:
```diff
@@ -0,0 +1,32 @@
+            {
+                EventPeriod eventPeriod = new EventPeriod(getEventQuery.StartEvent, getEventQuery.EndEvent);
+                Event foundEvent = await _eventRepository.GetEvent(getEventQuery.UserId, eventPeriod);
+                return new QueryResult<Event>(validationResult, foundEvent);
```
А нужно ли нам тут передавать результат валидации? Если по сути ошибок нет?

SherlockD commented on Application/Validation/ValidationResult.cs:
```diff
@@ -0,0 +1,26 @@
+{
+    public class ValidationResult
+    {
+        public bool IsFail { get; private set; } = false;
```
IsFail => Error == null, можно свойством сделать

Nafanyan committed:
1. Сделал свойство IsFail автосвойством
2. При успешном прохождении валидации теперь объект QueryResult теперь не будет содержать объект класса ValidationResult
3. Во всех файлах валидации ушел от переменной string error передавая текст ошибки в функцию Fail()

Nafanyan committed:
Небольшие правки с CreateEvent

SherlockD commented on Application/Users/Commands/UpdateUserLogin/UpdateUserLoginCommandHandler.cs:
```diff
@@ -0,0 +1,31 @@
+            if (!validationResult.IsFail)
+            {
+                User user = await _userRepository.GetById(updateUserLoginCommand.Id);
+                user.SetLogin(updateUserLoginCommand.Login);
```
А сохранение где? IUnitOfWirk.Commit

Nafanyan committed:
добавил unit of work и убрал дубликацю кода

Nafanyan committed:
создал предикаты для метода contains репозитория IUserRepository

Nafanyan committed:
Сделал предикаты для поиска по id

Nafanyan committed:
()
